crypto/serializer/deserializer_lib.c

```diff
@@ -0,0 +1,358 @@
+        return NULL;
+    return deser_inst->deser;
+}
+
+void *OSSL_DESERIALIZER_INSTANCE_deserializer_ctx
+    (OSSL_DESERIALIZER_INSTANCE *deser_inst)
+{
+    if (deser_inst == NULL)
+        return NULL;
+    return deser_inst->deserctx;
+}
+
+static int deser_process(const OSSL_PARAM params[], void *arg)
+{
+    struct deser_process_data_st *data = arg;
+    OSSL_DESERIALIZER_CTX *ctx = data->ctx;
+    OSSL_DESERIALIZER_INSTANCE *deser_inst = NULL;
+    OSSL_DESERIALIZER *deser = NULL;
+    BIO *bio = data->bio;
+    long loc;
+    size_t i;
+    int ok = 0;
+    /* For recursions */
+    struct deser_process_data_st new_data;
+
+    memset(&new_data, 0, sizeof(new_data));
+    new_data.ctx = data->ctx;
+
+    if (params == NULL) {
+        /* First iteration, where we prepare for what is to come */
+
+        data->current_deser_inst_index =
+            OSSL_DESERIALIZER_CTX_num_deserializers(ctx);
+
+        bio = data->bio;
+    } else {
+        const OSSL_PARAM *p;
+
+        deser_inst =
+            sk_OSSL_DESERIALIZER_INSTANCE_value(ctx->deser_insts,
+                                                data->current_deser_inst_index);
+        deser = OSSL_DESERIALIZER_INSTANCE_deserializer(deser_inst);
+
+        if (ctx->finalizer(deser_inst, params, ctx->finalize_arg)) {
+            ok = 1;
+            goto end;
+        }
+
+        /* The finalizer didn't return success */
+
+        /*
+         * so we try to use the object we got and feed it to any next
+         * deserializer that will take it.  Object references are not
+         * allowed for this.
+         * If this data isn't present, deserialization has failed.
+         */
+
+        p = OSSL_PARAM_locate_const(params, OSSL_DESERIALIZER_PARAM_DATA);
+        if (p == NULL || p->data_type != OSSL_PARAM_OCTET_STRING)
+            goto end;
+        new_data.bio = BIO_new_mem_buf(p->data, (int)p->data_size);
+        if (new_data.bio == NULL)
+            goto end;
+        bio = new_data.bio;
+    }
+
+    /*
+     * If we have no more deserializers to look through at this point,
+     * we failed
+     */
+    if (data->current_deser_inst_index == 0)
+        goto end;
+
+    if ((loc = BIO_tell(bio)) < 0) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_BIO_LIB);
+        goto end;
+    }
+
+    for (i = data->current_deser_inst_index; i-- > 0;) {
+        OSSL_DESERIALIZER_INSTANCE *new_deser_inst =
+            sk_OSSL_DESERIALIZER_INSTANCE_value(ctx->deser_insts, i);
+        OSSL_DESERIALIZER *new_deser =
+            OSSL_DESERIALIZER_INSTANCE_deserializer(new_deser_inst);
+
+        /*
+         * If |deser| is NULL, it means we've just started, and the caller
+         * may have specified what it expects the initial input to be.  If
+         * that's the case, we do this extra check.
+         */
+        if (deser == NULL && ctx->start_input_type != NULL
+            && strcmp(ctx->start_input_type, deser_inst->input_type) != 0)
```

Comment:
I cant see where you ever set start_input_type

Reply:
That's because I forgot to implement that.  Thanks for pointing it out.

Reply:
Done